File: frontend/mobile/lib/core/money.dart
import 'dart:math';

enum Currency { syp, usd, eur, try_ }

class Money {
  final int _fils;
  final Currency currency;

  Money(this._fils, {this.currency = Currency.syp}) {
    if (_fils < 0) {
      throw ArgumentError('المبلغ لا يمكن أن يكون سالباً');
    }
  }

  factory Money.fromFils(int fils, {Currency currency = Currency.syp}) {
    return Money(fils, currency: currency);
  }

  factory Money.fromSYP(num amount, {Currency currency = Currency.syp}) {
    final fils = (amount * 1000).round();
    return Money(fils, currency: currency);
  }

  int get fils => _fils;

  double get syp => _fils / 1000;

  Money operator +(Money other) {
    _assertSameCurrency(other);
    return Money(_fils + other._fils, currency: currency);
  }

  Money operator -(Money other) {
    _assertSameCurrency(other);
    return Money(_fils - other._fils, currency: currency);
  }

  Money operator *(num multiplier) {
    final fils = (_fils * multiplier).round();
    return Money(fils, currency: currency);
  }

  bool operator >(Money other) {
    _assertSameCurrency(other);
    return _fils > other._fils;
  }

  bool operator <(Money other) {
    _assertSameCurrency(other);
    return _fils < other._fils;
  }

  @override
  bool operator ==(Object other) =>
      other is Money && _fils == other._fils && currency == other.currency;

  @override
  int get hashCode => Object.hash(_fils, currency);

  String format() {
    final value = syp.toStringAsFixed(3);
    switch (currency) {
      case Currency.syp:
        return '$value ل.س';
      case Currency.usd:
        return '\$$value';
      case Currency.eur:
        return '$value €';
      case Currency.try_:
        return '$value ₺';
    }
  }

  void _assertSameCurrency(Money other) {
    if (currency != other.currency) {
      throw ArgumentError(
        'عملة غير متطابقة: متوقع ${currency.name}، مستلم ${other.currency.name}',
      );
    }
  }
}

File: frontend/admin/src/core/money.ts
export enum Currency {
  SYP = 'SYP',
  USD = 'USD',
  EUR = 'EUR',
  TRY = 'TRY',
}

export class Money {
  private readonly _fils: number;
  readonly currency: Currency;

  constructor(fils: number, currency: Currency = Currency.SYP) {
    if (fils < 0) throw new Error('المبلغ لا يمكن أن يكون سالباً');
    this._fils = fils;
    this.currency = currency;
  }

  static fromFils(fils: number, currency: Currency = Currency.SYP): Money {
    return new Money(fils, currency);
  }

  static fromSYP(amount: number, currency: Currency = Currency.SYP): Money {
    return new Money(Math.round(amount * 1000), currency);
  }

  get fils(): number {
    return this._fils;
  }

  get syp(): number {
    return this._fils / 1000;
  }

  add(other: Money): Money {
    this.assertSameCurrency(other);
    return new Money(this._fils + other._fils, this.currency);
  }

  subtract(other: Money): Money {
    this.assertSameCurrency(other);
    return new Money(this._fils - other._fils, this.currency);
  }

  multiply(multiplier: number): Money {
    return new Money(Math.round(this._fils * multiplier), this.currency);
  }

  isGreaterThan(other: Money): boolean {
    this.assertSameCurrency(other);
    return this._fils > other._fils;
  }

  isLessThan(other: Money): boolean {
    this.assertSameCurrency(other);
    return this._fils < other._fils;
  }

  equals(other: Money): boolean {
    return this._fils === other._fils && this.currency === other.currency;
  }

  format(): string {
    const value = this.syp.toLocaleString('ar-SA', {
      minimumFractionDigits: 3,
      maximumFractionDigits: 3,
    });
    const symbols: Record<Currency, string> = {
      [Currency.SYP]: 'ل.س',
      [Currency.USD]: '$',
      [Currency.EUR]: '€',
      [Currency.TRY]: '₺',
    };
    return `${value} ${symbols[this.currency]}`;
  }

  private assertSameCurrency(other: Money): void {
    if (this.currency !== other.currency) {
      throw new Error(
        `عملة غير متطابقة: متوقع ${this.currency}، مستلم ${other.currency}`,
      );
    }
  }
}
